src/web-extensions.c

```diff
@@ -0,0 +1,696 @@
+}
+
+static void
+web_extensions_session_class_init (WebExtensionsSessionClass *klass)
+{
+  GObjectClass *object_class;
+  SessionClass *session_class;
+
+  object_class = G_OBJECT_CLASS (klass);
+  object_class->finalize = web_extensions_session_finalize;
+
+  session_class = (SessionClass *)klass;
+  session_class->close = web_extensions_session_close;
+}
+
+static WebExtensionsSession *
+web_extensions_session_new (GVariant *options,
+                            Call *call,
+                            GDBusConnection *connection,
+                            GError **error)
+{
+  Session *session;
+  const char *session_token;
+
+  session_token = lookup_session_token (options);
+  session = g_initable_new (web_extensions_session_get_type (), NULL, error,
+                            "sender", call->sender,
+                            "app-id", xdp_app_info_get_id (call->app_info),
+                            "token", session_token,
+                            "connection", connection,
+                            NULL);
+
+  if (session)
+    g_debug ("webextensions session owned by '%s' created", session->sender);
+
+  return (WebExtensionsSession *)session;
+}
+
+static gboolean
+handle_create_session (XdpDbusWebExtensions *object,
+                       GDBusMethodInvocation *invocation,
+                       GVariant *arg_options)
+{
+  Call *call = call_from_invocation (invocation);
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  g_autoptr(GError) error = NULL;
+  Session *session;
+
+  session = (Session *)web_extensions_session_new (arg_options, call, connection, &error);
+  if (!session)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+  if (!session_export (session, &error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      session_close (session, FALSE);
+      return TRUE;
+    }
+  session_register (session);
+
+  xdp_dbus_web_extensions_complete_create_session (object, invocation, session->id);
+
+  return TRUE;
+}
+
+static void
+on_server_exited (GPid pid,
+                  gint status,
+                  gpointer user_data)
+{
+  Session *session = user_data;
+  WebExtensionsSession *we_session = (WebExtensionsSession *)session;
+
+  SESSION_AUTOLOCK (session);
+  we_session->child_pid = -1;
+  we_session->child_watch_id = 0;
+  session_close (session, TRUE);
+}
+
+static gboolean
+array_contains (JsonArray *array,
+                const char *value)
+{
+  guint length, i;
+
+  if (array == NULL)
+    return FALSE;
+
+  length = json_array_get_length (array);
+  for (i = 0; i < length; i++)
+    {
+      const char *element = json_array_get_string_element (array, i);
+      if (g_strcmp0 (element, value) == 0)
+        return TRUE;
+    }
+  return FALSE;
+}
+
+static gboolean
+is_valid_name (const char *name)
+{
+  return g_regex_match_simple ("^\\w+(\\.\\w+)*$", name, 0, 0);
```

Comment:
This is a bit obscure to me. Could you document in a comment what this regex is expected to accept and reject, and add tests exercising this validation?